### PR TITLE
Mssql provider fixes and optimisations

### PR DESCRIPTION
**Pull request by nyalldawson**
This adds a bunch of fixes and optimizations for the mssql provider:
- adds a provider test
- implements an expression compiler, for nice and fast expression based feature requests
- implements provider side feature IDs filtering, for nice and fast operations like zooming to selected or saving selected records
- better error handling when a query can't be successfully evaluated by the server (was getting crashes and other odd behaviour before this)

### CODE REVIEW

**Comment by NathanW2** [comment]
LGTM +1 to merge

File: src/core/qgssqlexpressioncompiler.h [hunk 1]
```diff
@@ -78,9 +78,11 @@ class CORE_EXPORT QgsSqlExpressionCompiler
 
     /** Returns a quoted attribute value, in the format expected by the provider.
      * Derived classes should override this if special handling of attribute values is required.
+     * @param value value to quote
+     * @param ok wil be set to true if value can be compiled
```
Comment (jef-n) [inline src/core/qgssqlexpressioncompiler.h]: typo

**Comment by nyalldawson** [comment]
Merged

**Comment by NathanW2** [comment]
Thanks for your great work.

**Pull request closed without merging**